net/openconnect/Makefile

```diff
@@ -34,7 +35,7 @@ endef
 define Package/openconnect
   SECTION:=net
   CATEGORY:=Network
-  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken
+  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken $(ICONV_DEPENDS) $(INTL_DEPENDS)
   TITLE:=OpenConnect VPN client (Cisco AnyConnect and Juniper/Pulse compatible)
```

Comment:
I do not think that this is the right solution. openconnect doesn't depend on iconv, so it shouldn't link to it. Since libxml2 depends in it, it should have done the right thing and added the right dependencies. That's what the rest of the libraries linked with like gnutls and openssl do.

Reply:
Funnily enough openconnect does actually (try to) use iconv:

```
Disabling Docbook support
Enabled Schematron support
checking iconv.h usability... yes
checking iconv.h presence... yes
checking for iconv.h... yes
checking for iconv... no
checking for iconv in -liconv... yes
checking for iconv declaration...
        extern size_t iconv (iconv_t cd,  char * *inbuf, size_t *inbytesleft, char * *outbuf, size_t *outbytesleft);
Disabling ICU support
```

It seems to use iconv from libc if available. But when you tell it that you want it to use external libiconv/libintl it will:

```
diff --git a/net/openconnect/Makefile b/net/openconnect/Makefile
index 59e81f77a..3fd616227 100644
--- a/net/openconnect/Makefile
+++ b/net/openconnect/Makefile
@@ -26,6 +26,7 @@ PKG_CONFIG_DEPENDS:= \
 PKG_USE_MIPS16:=0
 
 include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/nls.mk
 
 define Package/openconnect/config
        source "$(SOURCE)/Config.in"
@@ -55,6 +56,8 @@ endef
 
 CONFIGURE_ARGS += \
        --disable-shared \
+       --with-libiconv-prefix=$(ICONV_PREFIX) \
+       --with-libintl-prefix=$(INTL_PREFIX) \
        --with-vpnc-script=/lib/netifd/vpnc-script \
        --without-libpcsclite \
        --without-stoken \
```

readelf shows the openconnect binary now links to external libiconv/libintl (this is with CONFIG_BUILD_NLS=y).

```
 0x00000001 (NEEDED)                     Shared library: [libz.so.1]
 0x00000001 (NEEDED)                     Shared library: [libgnutls.so.30]
 0x00000001 (NEEDED)                     Shared library: [libxml2.so.2]
 0x00000001 (NEEDED)                     Shared library: [libintl.so.8]
 0x00000001 (NEEDED)                     Shared library: [libiconv.so.2]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
 0x0000000f (RPATH)                      Library rpath: [/home/sk/tmp/openwrt/staging_dir/target-mips_24kc_musl/usr/lib/libintl-full/lib:/home/sk/tmp/openwrt/staging_dir/target-mips_24kc_musl/usr/lib/libiconv-full/lib]
```

Functionality-wise as an OpenWrt user I would expect it to follow the OpenWrt NLS configuration (in menuconfig). So if NLS is off you get iconv stub without NLS support and with NLS on you get everything via libiconv and libintl. openconnect using iconv from libc (musl/glibc) I would not expect, also because musl iconv is likely rather limited.

To sum things up, I'd suggest to use the change proposed by @mhei and add the configure switches on top.

Kind regards,
Seb

Reply:
@nmav: 

> openconnect doesn't depend on iconv or libintl, so it shouldn't link to it. Since libxml2 depends in it,
> it should have done the right thing and added the right dependencies.
> That's what the rest of the libraries linked with like gnutls and openssl do.

I fully agree that in an ideal world, the API of libxml2 would be so abstracted that library users do need to bother whether libxml2 is iconv/intl capable or compiled against whatever other library. But this seems to be not the case...

I admit, that I missed the CONFIGURE_ARGS stuff since I did not run it by hand, I just reproduced the compile error and the absence of it with my proposal. I can easily add these lines and repush to this PR.
What do you think?

Reply:
But isn't the libxml2.pc file mentioning these libraries? If it is, there shouldn't be any need to explicitly link the dependencies as they would get these from pkg-config.

Reply:
> Funnily enough openconnect does actually (try to) use iconv:
[...]

True, but there is little need to have openconnect with NLS in openwrt, at least no-one has asked for it.

> Functionality-wise as an OpenWrt user I would expect it to follow the OpenWrt NLS configuration (in menuconfig). So if NLS is off you get iconv stub without NLS support and with NLS on you get everything via libiconv and libintl. openconnect using iconv from libc (musl/glibc) I would not expect, also because musl iconv is likely rather limited.
> 
> To sum things up, I'd suggest to use the change proposed by @mhei and add the configure switches on top.

I still think that the libxml2.pc should contain the right libraries in order for openconnect to be able to link with libxml2. That way we can decide whether to bring libintl to openconnect independently, if needed, not just because libxml2 needs it.

Reply:
@mhei would it be possible to add libiconv into libxml2.pc? In that case doesn't it solve the compilation issue?

Reply:
"-liconv" is already in the pc file. But to find the correct path to the lib that libxml2 was built against (be it iconv stub or libiconv-full) while building openconnect requires including nls.mk, because this include adds the correct paths depending on CONFIG_BUILD_NLS.